[FLINK-2583] Add Stream Sink For Rolling HDFS Files

aljoscha opened the pull request:
Note: The rolling sink is not yet integrated with checkpointing/fault-tolerance.

rmetzger commented on flink-staging/flink-streaming/flink-streaming-connectors/flink-connector-hdfs/pom.xml:
```diff
@@ -0,0 +1,107 @@
+			<version>${project.version}</version>
+		</dependency>
+
+		<dependency>
```
why are you not using our shaded hadoop dependency?

StephanEwen commented:
Looks very good from a first glance!

Can you explain how the writing behaves in cases of failues? Will it start a new file? Will it try to append to the previous one? What about incomplete records?

aljoscha commented:
If it fails in the middle of writing or before sync/flush is called on the writer then the data can be in an inconsistent state. I see three ways of dealing with this, one is more long-term.

The long term solution is to make the sink exactly-once aware. Either using truncate() support in Hadoop 2.7 or a custom Thread that does merging of part files and throwing away of data that was erroneously written.

The two short term options are:
 - Keep it as it is, consumers need to be able to deal with corrupt records and ignore them. This would give you at-least-once semantics.
 - Write to a temporary file. When rolling, close the current bucket and rename the file to the final filename. This would ensure that the output doesn't contain corrupt records but you would have neither at-least-once nor exactly-once semantics because some written records would be lost if checkpoint restore restores to a state after the writing of the current bucket file started.

tillrohrmann commented:
Concerning the second short term option: Is the problem there that the checkpointing is not aligned with the rolling? Thus, you can take a checkpoint but you still have a temporary file which contains all the records up to the checkpoint. If the sink now fails before renaming the temporary file, then the records which have been written before the checkpoint was taken are lost. Did I understand it correctly? If this is the case, then the records in the temporary file should actually be part of the state of the sink, right?

aljoscha commented:
Should be, but then we can also just keep them in memory and write when checkpointing. But this has more potential of blowing up because OOM.

tillrohrmann commented:
What about renaming the files upon checkpointing? I mean not triggering the rolling mechanism but to write the current temp file as the latest part and start a new part after the checkpoint. Would this help? Of course this would entail that the different file parts will have different sizes.

rmetzger commented:
I'm currently trying out the module. Some comments:
- Why do we name the module `flink-connector-hdfs`. I think a name such as `flink-connector-filesystems` or `flink-connector-hfs` would be more appropriate because its is implemented against Hadoops `FileSystems` classes, and not HDFS classes.
So users should be able to use the connector with other FS such as Tachyon, NFS, S3 etc.

- Is there a way of re-using existing InputFormats with the rolling file sink? I guess users will start asking about CSV, Avro, Parquet ...

- I think there is already some code to monitor a file system directory to ingest it into a data stream. Maybe it would make sense to move that code out of the core into this module?

rmetzger commented on docs/apis/streaming_guide.md:
```diff
@@ -1836,6 +1837,110 @@ More about information about Elasticsearch can be found [here](https://elastic.c
+The only required parameter is the base path in HDFS where the rolling files (buckets) will be
+stored. The sink can be configured by specifying a custom bucketer, HDFS writer and batch size.
+
+By default the rolling sink will use the pattern `"yyyy-MM-dd--HH"` to name the rolling buckets.
```
Can you make it a bit more explicit that a new directory is created when the pattern changes?

StephanEwen commented:
I think using truncate for exactly once is the way to go. To support users with older HDFS versions, how about this:

1. We consider only valid what was written successfully at a checkpoint (hflush/hsync). When we roll over to a new file on restart, we write a `.length` file for that other file that indicates how many bytes are valid in that file. Basically simulating truncate by adding a metadata file.

2. Optionally, the user can activate a merge-on roll-over, that takes all the files from the attempts and all the metadata files, and merges them into one file. This rollover can be written such that it works incrementally and re-tries on failures, etc...

aljoscha commented:
Question is, should we do exactly once now or put it in as it is (more or less)?

aljoscha commented:
I (almost) completely reworked the sink. It is now called `RollingSink` and the module is called `flink-connector-filesystem` to show that it works with any Hadoop FileSystem.

It is integrated with the checkpointing mechanism to provide exactly-once semantics. When supported it will use `truncate` for this. Otherwise it will write a special `.valid-length` file that specifies how many bytes in a file are valid.

I added an ITCase that verifies the exactly-once behavior.

I added a lot of description about how the sink works in the Javadoc of `RollingSink`, so if you want to check it out I suggest you start there.

StephanEwen commented:
How hard is it to support the truncate file code path also for regular Unix file systems (rather than only HDFS 2.7+)?

The reason is that this way we would support mounted file systems (NAS, EBS) and MapR file system (exposes also a regular Unix file system interface) in a seamless exactly once way (no need to have readers that are aware of the metadata files).

aljoscha commented:
What do you mean? This is using the Hadoop FileSystem for everything. Is your suggestion to abstract away the filesystems behind our own FileSystem class again?

By the way, truncate works with "file://", for example. I didn't check other filesystems.

StephanEwen commented:
I mean can we get the "truncate()" behavior for local file systems even if you build for Hadoop versions earlier than  2.7?

aljoscha commented:
It could be done by circumventing FileSystem and just using the regular Java File I/O API to perform the truncate if we detect that the FileSystem works on file "file://" paths.

Should we do this? It seems like a bit of a hack to me.

StephanEwen commented:
Let's add this after we merge this, but it sounds quite valueable to me...

aminouvic commented:
Since the RollingSink was a little inspired by flume's HDFS Sink, it would be nice to include another really valuable features that could make it more complete.

One of the most common use cases of the HDFS Sink is to dispatch data into multiple directories depending of attributes present in source events.

For example, let's say we have some data that have a timestamp and status fields, one can specify in Flume conf file to write data into different directories like this:
hadfs.path=/somepath/%{timestamp}/%{status}
The result will be to write data into multiple folders like 
/somepath/some_timestamp/wellformed
/somepath/some_timestamp/malformed
/somepath/some_timestamp/incomplete 
... 
etc

To achieve this, Flume maintains a LRU Hashmap to hold not one but a set of Writers, and computes the destination path for each event.
It also uses some params like maxOpenfFiles, idleTimeOut(optional, used to close a file after x seconds of inactivity) to ensure not having too many open files.

But to include this, the bucketing, cleanOnStartup and the checkpointing logics need to be changed.

fhueske commented:
Yes, we just had a discussion on the user mailing list about a partitioned output format for batch processing. Definitely a good addition for both DataSet and DataStream API.

How about we merge this PR (once it is ready) and add the partitioning functionality later?

aljoscha commented:
Yes, I would be very much in favor of adding it in the current incarnation before adding more stuff.

It sill has a bug right now with exactly once, somehow some elements seem to get lost. I'm still debugging.

StephanEwen commented:
Yes, let's merge it before extending it.

@aminouvic If you want, can you create a JIRA with a description of that behavior, as a followup task?

As a workaround, you should be able to do this right now with the help of splitting data streams and having multiple HDFS sinks.

aminouvic commented:
Yeah you're right better have an operational version of the sink first, followup JIRA created https://issues.apache.org/jira/browse/FLINK-2672

aljoscha commented:
I fixed the bug and now I'm confident that it should work.

Also, I updated the travis build matrix because the sink does not work with hadoop 2.0.0-alpha. We had discussions about changing it, so I hope it is OK.

rmetzger commented on .travis.yml:
```diff
@@ -19,9 +19,9 @@ matrix:
-      env: PROFILE="-P!include-yarn -Dhadoop.version=2.0.0-alpha"
-    - jdk: "oraclejdk7"
-      env: PROFILE="-Dhadoop.version=2.2.0"
+      env: PROFILE="-Dhadoop.version=2.4.0"
```
As per mailing list discussion, we agreed to support hadoop 2.3.0.
But it seems that the yarn tests are not working with 2.3.0, so if you want we can set the version to 2.3.0 when the tests are fixed.

aljoscha replied:
So you are OK with merging it as is and then changing the build matrix later?

rmetzger replied:
I am okay, but this is a pretty important thing. Maybe we should wait for at least one more committer agreeing with us.

rmetzger commented:
I think the pull request has grown quite a lot. I think we should merge it now and then improve it from there.

StephanEwen commented:
Looks good, will merge this!
